utcp.c

```diff
@@ -509,15 +627,157 @@ static void retransmit(struct utcp_connection *c) {
 		case FIN_WAIT_2:
 			// We shouldn't need to retransmit anything in this state.
 #ifdef UTCP_DEBUG
+			debug("Error: retransmit unexpected connection state %p %s\n", c, strstate[c->state]);
 			abort();
 #endif
-			timerclear(&c->rtrx_timeout);
-			break;
+			stop_retransmit_timer(c);
+			goto cleanup;
 	}
 
+	start_retransmit_timer(c);
+	utcp->rto *= 2;
+	if(utcp->rto > MAX_RTO)
+		utcp->rto = MAX_RTO;
+	c->rtt_start.tv_sec = 0; // invalidate RTT timer
+
+cleanup:
 	free(pkt);
 }
 
+/* Update receive buffer and SACK entries after consuming data.
+ *
+ * Situation:
+ *
+ * |.....0000..1111111111.....22222......3333|
+ * |---------------^
+ *
+ * 0..3 represent the SACK entries. The ^ indicates up to which point we want
+ * to remove data from the receive buffer. The idea is to substract "len"
+ * from the offset of all the SACK entries, and then remove/cut down entries
+ * that are shifted to before the start of the receive buffer.
+ *
+ * There are three cases:
+ * - the SACK entry is ahead of ^, in that case just change the offset.
+ * - the SACK entry starts before and ends after ^, so we have to
+ *   change both its offset and size.
+ * - the SACK entry is completely before ^, in that case delete it.
+ */
+static void sack_consume(struct utcp_connection *c, size_t len) {
+	debug("sack_consume %lu\n", (unsigned long)len);
+
+	buffer_get(&c->rcvbuf, NULL, len);
+
+	for(int i = 0; i < NSACKS && c->sacks[i].len; ) {
+		if(len < c->sacks[i].offset) {
+			c->sacks[i].offset -= len;
+			i++;
+		} else if(len < c->sacks[i].offset + c->sacks[i].len) {
+			c->sacks[i].len -= len - c->sacks[i].offset;
+			c->sacks[i].offset = 0;
+			i++;
+		} else {
+			if(i < NSACKS - 1) {
+				// move remaining sacks one ahead
+				memmove(&c->sacks[i], &c->sacks[i + 1], ((NSACKS - 1) - i) * sizeof c->sacks[i]);
+				// clean last sack len
+				c->sacks[NSACKS - 1].len = 0;
```

Comment:
I thought it might be just much easier to understand if this branch did something like
```
for ( int j = i; j < NSACKS; ++j )
    c->sacks[j].len = 0;
break;
```
Wouldn't that have the same outcome?